### ISSUE TITLE
A question about the weights of loss.

### ISSUE THREAD

**Issue by hfutzzw** [TomHeaven/msba_reid#3]
In the implementation of your code, the loss weights were directly multiplied by logits vector, rather than loss scalar.  The weighted logits will be normlized by softmax opt and the weights will not work.
![image](https://user-images.githubusercontent.com/30337787/89652971-e5f8d380-d8f8-11ea-9c3d-a1db3b7633b9.png)

**Comment by TomHeaven** [comment]
It's a good question. After doing some math, I find adding weights here is not strictly equivalent to adding weights to loss scalar. However, it does affects the total loss scalar.

Having result vector $x = [x_i ]$ and target vector $t = [t_i]$, log_softmax of $x$ is
$$
log_softmax(x_i) = log( exp(x_i) / sum_j (exp(x_j)) ) = x_i - log(sum_j(exp(x_j)))
$$
and the loss scalar used in the project is 
$$
  sum_i t_i * log_softmax(x_i) = sum_i t_i*( x_i - log(sum_j(exp(x_j))))
= sum_i  t_i * x_i - t_i * log(sum_j (exp( x_j))))
Now we multiply $x_i$ with a constant scalar $l$, we get
$$
   sum_i t_i * log_softmax(l x_i) = sum_i t_i*(l *x_i - log(sum_j(exp( l*x_j))))
= sum_i l * t_i* x_i - t_i * log(sum_j(exp( l *x_j))))
$$

If we weight loss scalar directly, we get
$$
  sum_i l * t_i * log_softmax(l x_i) 
=  sum_i l * t_i* x_i - t_i * l * log(sum_j(exp(x_j))))
$$

The first term is weighted by $l$ as expected, but that does not go for the second term since
$$
log(sum_j(exp( l *x_j)))) <>  l * log(sum_j(exp(x_j))))
$$

So there are actually some differences. I'll change the implementation here and do some experiments to evaluate the influence.

**Comment by chenxiangzZ** [comment]
there are six CE Loss ,why only four in your implementation?
![Uploading image.png…]()

**Comment by TomHeaven** [comment]
@chenxiangzZ 
Try adding the following code before `return`
```
print('len(lf_xent)', len(lf_xent))
```
and you will find the truth.
